### ISSUE TITLE
Upgrade to latest version failure

### ISSUE THREAD

**Issue by oferz-everc** [grokability/snipe-it#8933]
Thanks in advance for the assistance.
All details are listed below, if I've missed something please let me know and I'll add.
I'm trying to upgrade from 4.6.13 using upgrade.php and got few errors in output (pasted below)

#### Please confirm you have done the following before posting your bug report:

- [ ] I have enabled debug mode 
- [x] I have read [checked the Common Issues page](https://snipe-it.readme.io/docs/common-issues)

**Describe the bug**
Trying to upgrade from v4.6.13 to latest version using upgrade.php while getting an error.

**To Reproduce**
Running php upgrade.php from the snipe-it directory. 

**Expected behavior**
Expected a smooth upgrade to last version of snipe-it

**Screenshots**

**Server (please complete the following information):**
 - Snipe-IT Version v4.6.13
 - OS: Ubuntu 16.04
 - Web Server: Apache 2.4.18
 - PHP Version 7.0.33

**Error Messages**
`Welcome to the Snipe-IT upgrader.

Please note that this script will not download the latest Snipe-IT
files for you unless you have git installed.
It simply runs the standard composer and artisan
commands needed to finalize the upgrade after.

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! WARNING !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!! If you have any encrypted custom fields, BE SURE TO run the recrypter if upgrading from v3 to v4.
!! See the Snipe-IT documentation for help:
!! https://snipe-it.readme.io/docs/upgrading-to-v4
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! WARNING !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
--------------------------------------------------------
STEP 1: Backing up database:
--------------------------------------------------------

-- Starting backup...
Dumping database snipeit_db...
Backup failed because The dump process failed with exitcode 2 : Misuse of shell builtins : sh: 1: cannot create /var/www/snipeit.evercompliant.com/storage/laravel-backups/temp//snipeit_db.sql: Directory nonexistent
.
Backup failed because: The stream or file "/var/www/snipeit.evercompliant.com/storage/logs/laravel.log" could not be opened: failed to open stream: Permission denied.

In StreamHandler.php line 107:

  The stream or file "/var/www/snipeit.evercompliant.com/storage/logs/laravel
  .log" could not be opened: failed to open stream: Permission denied

--------------------------------------------------------
STEP 2: Putting application into maintenance mode:
--------------------------------------------------------

--
In StreamHandler.php line 107:

  The stream or file "/var/www/snipeit.evercompliant.com/storage/logs/laravel
  .log" could not be opened: failed to open stream: Permission denied

--------------------------------------------------------
STEP 3: Pulling latest from Git (master branch):
--------------------------------------------------------

Git is installed.
fatal: Unable to create '/var/www/snipeit.evercompliant.com/.git/index.lock': Permission denied
fatal: Unable to create '/var/www/snipeit.evercompliant.com/.git/index.lock': Permission denied
Cannot save the current index state
error: cannot open .git/FETCH_HEAD: Permission denied

-- -- -- --------------------------------------------------------
Step 4: Cleaning up old cached files:
--------------------------------------------------------

-- No bootstrap/cache/compiled.php, so nothing to delete.
-- Deleting bootstrap/cache/services.php. It it no longer used.
-- No bootstrap/cache/config.php, so nothing to delete.
-- Configuration cache cleared!
-- Cache cleared successfully.
-- Route cache cleared!
-- Compiled views cleared!

--------------------------------------------------------
Step 5: Updating composer dependencies:
(This may take an moment.)
--------------------------------------------------------

-- We couldn't find a local composer.phar - trying globally.

Generating optimized autoload files

  [ErrorException]
  file_put_contents(/var/www/snipeit.evercompliant.com/vendor/composer/autolo
  ad_namespaces.php): failed to open stream: Permission denied

dump-autoload [--no-scripts] [-o|--optimize] [-a|--classmap-authoritative] [--apcu] [--no-dev]

Loading composer repositories with package information
Installing dependencies from lock file
Nothing to install or update

  [ErrorException]
  file_put_contents(/var/www/snipeit.evercompliant.com/vendor/composer/instal
  led.json): failed to open stream: Permission denied

install [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [--no-suggest] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--apcu-autoloader] [--ignore-platform-reqs] [--] [<packages>]...

--------------------------------------------------------
Step 6: Migrating database:
--------------------------------------------------------

-- Nothing to migrate.

--------------------------------------------------------
Step 7: Checking for OAuth keys:
--------------------------------------------------------

- OAuth keys detected. Skipping passport install.

--------------------------------------------------------
Step 8: Caching routes and config:
--------------------------------------------------------

-- Configuration cache cleared!

In StreamHandler.php line 107:

  The stream or file "/var/www/snipeit.evercompliant.com/storage/logs/laravel
  .log" could not be opened: failed to open stream: Permission denied

-- Route cache cleared!

In StreamHandler.php line 107:

  The stream or file "/var/www/snipeit.evercompliant.com/storage/logs/laravel
  .log" could not be opened: failed to open stream: Permission denied

--------------------------------------------------------
Step 9: Taking application out of maintenance mode:
--------------------------------------------------------

-- Application is now live.

--------------------------------------------------------
FINISHED! Clear your browser cookies and re-login to use :
your upgraded Snipe-IT.
--------------------------------------------------------`

**Additional context**
- Is this a fresh install or an upgrade? - upgrade
- What OS and web server you're running Snipe-IT on - apache & ubuntu 16.04
- What method you used to install Snipe-IT (install.sh, manual installation, docker, etc) - git

**Comment by snipe** [comment]
> Backup failed because: The stream or file "/var/www/snipeit.evercompliant.com/storage/logs/laravel.log" could not be opened: failed to open stream: Permission denied.

According to that error message, the `storage` directory - or at the very least, the `storage/logs/laravel.log` file - cannot be written to. You'll need to fix that before proceeding.

**Comment by oferz-everc** [comment]
I've just double checked that.
The storage folder permissions are 775 for the www-data user/group, and recursively to laravel.log file.
The user I'm using is a member of www-data group, so there shouldn't be any issue there.
Or I'm missing something?

drwxrwxr-x  8 www-data www-data   4096 Apr  7  2019 storage
drwxrwxr-x 2 www-data www-data 4096 Mar 11  2019 logs
-rwxrwxr-x 1 www-data www-data 1902491 Apr 19  2020 laravel.log

**Comment by oferz-everc** [comment]
OK, I've rebooted the machine and fixed another permissions issue with the .git & vendor folders that was wrong.

I ran the upgrade.php again, but ended up not upgrading to latest version, with another permissions issue - this time missing permission to the "app" folder.
So I've fixed that, and had an issue this time with the "resources" folder - fixed that as well.
After running the upgrade again, Snipe-IT won't load - "Whoops, looks like something went wrong."
The output I've had this time attached as a file (upgrade-test.log), and the console output in a 2nd file (upgrade-test-console)
[upgrade-test.log](https://github.com/snipe/snipe-it/files/5729963/upgrade-test.log)
[upgrade-test-console.log](https://github.com/snipe/snipe-it/files/5729966/upgrade-test-console.log)

Fixed the permissions there, and checked again if the upgrade will work.
output files for the 2nd test:
[upgrade-test2.log](https://github.com/snipe/snipe-it/files/5730008/upgrade-test2.log)
[upgrade-test2-console.log](https://github.com/snipe/snipe-it/files/5730009/upgrade-test2-console.log)

I've seen the issue with being 1710 commits behind master, and found ticket #7215 - tried git pull & stash, but did not worked.
As stash seemed to be worked ok, still got error on pull regrading overwritten files.
I've tried "git reset --hard" - output was "HEAD is now at 6d98bd6 Fixed error if item requested or request was deleted (#6786)"

I've found [this StackOverFlow discussion](https://stackoverflow.com/questions/1125968/how-do-i-force-git-pull-to-overwrite-local-files) 
And followed the comment marked as solution.

Now, found that I'm missing two php extensions - installed both, and re-ran the upgrade.
Outputs:
[upgrade-test3.log](https://github.com/snipe/snipe-it/files/5730128/upgrade-test3.log)
[upgrade-test3-console.log](https://github.com/snipe/snipe-it/files/5730129/upgrade-test3-console.log)

And not the site is not reachable at all - error 500.

How can I continue from now on?
And, is there any way I could've avoided those issues? (I'm not the one who originally installed it)


Thanks for the help!

**Comment by oferz-everc** [comment]
Tried rebooting the machine, snipe-it site reached, but showing an error:
Composer detected issues in your platform: Your Composer dependencies require a PHP version ">= 7.2.0".

While php version is 7.4.13

**Comment by oferz-everc** [comment]
OK, Managed to solve the issue.

I've rolled back using a machine-level backup I've created.

I'll sum up every resolution step in case someone will see it in the future:

1. Double check current user permissions on the Snipe-IT folder (for me, after adding the user at first had to reboot the machine for it to be applied - ubuntu 16.04)
2. Upgrade to php 7.4 (latest) before running upgrade.php
3. git issue (commits gap) was resolved by running the steps in [this link](https://stackoverflow.com/questions/1125968/how-do-i-force-git-pull-to-overwrite-local-files)
4. only now run the upgrade and it works (for me)

And of course, always keep a backup right before doing stuffs like that :)

I'll close the ticket now.
Thanks!
